File: inter/byosnap-node-ts/src/controllers/usersController.ts
import { ErrorResponse, SuccessResponse } from '../types/responses';
import { ProfilePayload } from '../models/profilePayloadModel';
import { authMiddleware } from '../middleware/authMiddleware';
import { Request as ExpressRequest } from 'express';
import { Controller, Route, Get, Path, Post, Put, Delete, Extension, Body, Middlewares, TsoaResponse, Response, Res, Request } from 'tsoa';
import { ProfilesServiceApi }  from '../snapser-internal/api/profilesServiceApi'
import { UpsertProfileRequest } from '../snapser-internal/model/upsertProfileRequest';

// # @GOTCHAS 👋 - Please read GOTCHAS.md

@Route('/v1/byosnap-inter/users')
export class UserController extends Controller {

    /**
     * @summary Game APIs
     */
    @Get("{userId}/game")
    @Extension("x-description", 'This API will work with User and Api-Key auth. With a valid user token and api-key, you can access this API.')
    @Extension("x-snapser-auth-types", ["user", "api-key", "internal"])
    @Response<SuccessResponse>(200, "Successful Response")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["user", "api-key", "internal"])])
    public async getGame(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Path() userId: string,
        @Request() req: ExpressRequest,
    ): Promise<SuccessResponse> {
      const expressReq = req as ExpressRequest;
      const authType = expressReq.header("Auth-Type");
      const headerUserId = expressReq.header("User-Id");
      return {
        api: 'getGame',
        authType: authType ?? 'N/A',
        headerUserId: headerUserId ?? 'N/A',
        pathUserId: userId,
        message: 'success'
      };
    }

    /**
     * @summary Game APIs
     */
    @Post("{userId}/game")
    @Extension("x-description", 'This API will work only with Api-Key auth. You can access this API with a valid api-key.')
    @Extension("x-snapser-auth-types", ["api-key", "internal"])
    @Response<SuccessResponse>(200, "Successful Response")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["api-key", "internal"])])
    public async saveGame(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Path() userId: string,
        @Request() req: ExpressRequest,
    ): Promise<SuccessResponse> {
      const expressReq = req as ExpressRequest;
      const authType = expressReq.header("Auth-Type");
      const headerUserId = expressReq.header("User-Id");
      return {
        api: 'saveGame',
        authType: authType ?? 'N/A',
        headerUserId: headerUserId ?? 'N/A',
        pathUserId: userId,
        message: 'success'
      };
    }

    /**
     * @summary User APIs
     */
    @Delete("{userId}")
    @Extension("x-description", 'This API will work only when the call is coming from within the Snapend.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Response<SuccessResponse>(200, "Successful Response")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["internal"])])
    public async deleteUser(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Path() userId: string,
        @Request() req: ExpressRequest,
    ): Promise<SuccessResponse> {
      const expressReq = req as ExpressRequest;
      const gatewayHeader = expressReq.header("Gateway");
      const headerUserId = expressReq.header("User-Id");
      return {
        api: 'deleteUser',
        authType: gatewayHeader ?? 'N/A',
        headerUserId: headerUserId ?? 'N/A',
        pathUserId: userId,
        message: 'success'
      };
    }

    /**
     * @summary User APIs
     */
    @Put("{userId}/profile")
    @Extension("x-description", 'This API will work for all auth types.')
    @Extension("x-snapser-auth-types", ["user", "api-key", "internal"])
    @Response<SuccessResponse>(200, "Successful Response")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Response<ErrorResponse>(400, "Bad Request")
    @Middlewares([authMiddleware(["user", "api-key", "internal"])])
    public async updateUserProfile(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Res() _badRequest: TsoaResponse<400, ErrorResponse>,
        @Path() userId: string,
        @Request() req: ExpressRequest,
        @Body() body: ProfilePayload
    ): Promise<SuccessResponse> {
      const expressReq = req as ExpressRequest;
      const authType = expressReq.header("Auth-Type");
      const headerUserId = expressReq.header("User-Id");
      const baseUrl = process.env.SNAPEND_PROFILES_HTTP_URL ?? 'http://profiles-service:8090';
      const profilesApi = new ProfilesServiceApi(baseUrl);
      const payload: UpsertProfileRequest = {
        profile: body.profile
      };
      // TODO: Uncomment the following code
      // try {
      //   const result = await profilesApi.profilesInternalUpsertProfile(userId, 'internal', payload);
      //   const body = result.body;

      //   return {
      //     api: 'updateUserProfile',
      //     authType: authType ?? 'N/A',
      //     headerUserId: headerUserId ?? 'N/A',
      //     pathUserId: userId,
      //     message: JSON.stringify(body)
      //   };
      // } catch (error) {
      //     //Send ErrorResponse
      //     return _badRequest(400, {
      //       error_message: error?.message || "Upsert failed"
      //     });
      // }
      // TODO: Once you uncomment the above code, remove the following code
      return {
        api: 'updateUserProfile',
        authType: authType ?? 'N/A',
        headerUserId: headerUserId ?? 'N/A',
        pathUserId: userId,
        message: 'Remove this'
      };
    }
}

File: inter/ByoSnapCSharp/Controllers/UsersController.cs
using Microsoft.AspNetCore.Mvc;
using ByoSnapCSharp.Filters;
using ByoSnapCSharp.Models;
using ByoSnapCSharp.Utilities;
using System.Linq;
using Microsoft.OpenApi.Models;
using Microsoft.OpenApi.Any;
using Swashbuckle.AspNetCore.Annotations;
using System;
using SnapserInternal.Api;
using SnapserInternal.Client;
using SnapserInternal.Model;
using System.Net.Http;


namespace ByoSnapCSharp.Controllers
{
  [ApiController]
  [Route("v1/byosnap-inter/users/{UserId}")]
  [Produces("application/json")]
  public class UsersController : ControllerBase
  {
    [HttpGet("game")]
    [SnapserAuth(AppConstants.userAuthType, AppConstants.apiKeyAuthType, AppConstants.internalAuthType)]
    [ValidateAuthorization(AppConstants.userAuthType, AppConstants.apiKeyAuthType, AppConstants.internalAuthType)]
    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(SuccessResponseSchema))]
    [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(ErrorResponseSchema))]
    [ProducesResponseType(StatusCodes.Status401Unauthorized, Type = typeof(ErrorResponseSchema))]
    [SwaggerOperation(OperationId = "GetGame", Summary = "Game APIs", Description = "This API will work with User and Api-Key auth. With a valid user token and api-key, you can access this API.")]
    public ActionResult<SuccessResponseSchema> GetGame([FromRoute] UserIdParameterSchema userParams)
    {
      var authTypeHeader = HttpContext.Request.Headers[AppConstants.authTypeHeaderKey].FirstOrDefault();
      var userIdHeader = HttpContext.Request.Headers[AppConstants.userIdHeaderKey].FirstOrDefault();
      return Ok(new SuccessResponseSchema
      {
        Api = "GetGame",
        AuthType = authTypeHeader ?? "N/A",
        HeaderUserId = userIdHeader ?? "N/A",
        PathUserId = userParams.UserId,
        Message = "success"
      });
    }

    [HttpPost("game")]
    [SnapserAuth(AppConstants.apiKeyAuthType, AppConstants.internalAuthType)]
    [ValidateAuthorization(AppConstants.apiKeyAuthType, AppConstants.internalAuthType)]
    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(SuccessResponseSchema))]
    [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(ErrorResponseSchema))]
    [ProducesResponseType(StatusCodes.Status401Unauthorized, Type = typeof(ErrorResponseSchema))]
    [SwaggerOperation(OperationId = "SaveGame", Summary = "Game APIs", Description = "This API will work only with Api-Key auth. You can access this API with a valid api-key.")]
    public ActionResult<SuccessResponseSchema> SaveGame([FromRoute] UserIdParameterSchema userParams)
    {
      var authTypeHeader = HttpContext.Request.Headers[AppConstants.authTypeHeaderKey].FirstOrDefault();
      var userIdHeader = HttpContext.Request.Headers[AppConstants.userIdHeaderKey].FirstOrDefault();
      return Ok(new SuccessResponseSchema
      {
        Api = "SaveGame",
        AuthType = authTypeHeader ?? "N/A",
        HeaderUserId = userIdHeader ?? "N/A",
        PathUserId = userParams.UserId,
        Message = "success"
      });
    }

    [HttpDelete("")]
    [SnapserAuth(AppConstants.internalAuthType)]
    [ValidateAuthorization(AppConstants.internalAuthType)]
    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(SuccessResponseSchema))]
    [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(ErrorResponseSchema))]
    [ProducesResponseType(StatusCodes.Status401Unauthorized, Type = typeof(ErrorResponseSchema))]
    [SwaggerOperation(OperationId = "DeleteUser", Summary = "User APIs", Description = "This API will work only when the call is coming from within the Snapend.")]
    public ActionResult<SuccessResponseSchema> DeleteUser([FromRoute] UserIdParameterSchema userParams)
    {
      var gatewayHeader = HttpContext.Request.Headers[AppConstants.gatewayHeaderKey].FirstOrDefault();
      var userIdHeader = HttpContext.Request.Headers[AppConstants.userIdHeaderKey].FirstOrDefault();
      return Ok(new SuccessResponseSchema
      {
        Api = "DeleteUser",
        AuthType = gatewayHeader ?? "N/A",
        HeaderUserId = userIdHeader ?? "N/A",
        PathUserId = userParams.UserId,
        Message = "success"
      });
    }

    [HttpPut("profile")]
    [SnapserAuth(AppConstants.userAuthType, AppConstants.apiKeyAuthType, AppConstants.internalAuthType)]
    [ValidateAuthorization(AppConstants.userAuthType, AppConstants.apiKeyAuthType, AppConstants.internalAuthType)]
    [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(SuccessResponseSchema))]
    [ProducesResponseType(StatusCodes.Status400BadRequest, Type = typeof(ErrorResponseSchema))]
    [ProducesResponseType(StatusCodes.Status401Unauthorized, Type = typeof(ErrorResponseSchema))]
    [SwaggerOperation(OperationId = "UpdateUserProfile", Summary = "User APIs", Description = "This API will work for all auth types.")]
    public ActionResult<SuccessResponseSchema> UpdateUserProfile(
      [FromRoute] UserIdParameterSchema userParams,
      [FromBody] UpsertProfileRequestWrapper body)
    {
      try
      {
        var gatewayHeader = HttpContext.Request.Headers[AppConstants.gatewayHeaderKey].FirstOrDefault();
        var userIdHeader = HttpContext.Request.Headers[AppConstants.userIdHeaderKey].FirstOrDefault();
        Object result = null;
        // TODO: Uncomment the following code to use the SnapserInternal API
        // // // 👇 [IMPORTANT]: You have to set the baseURL using the Snapser provided
        // // //    environment variables. There will be a different environment variable
        // // //    for each snap. Eg: for the Profile Snap, Snapser sets the
        // // //    SNAPEND_PROFILES_HTTP_URL, For the Auth snap it will be SNAPEND_AUTH_HTTP_URL
        // Configuration config = new Configuration();
        // config.BasePath = Environment.GetEnvironmentVariable("SNAPEND_PROFILES_HTTP_URL") ?? config.GetOperationServerUrl("ProfilesServiceApi.ProfilesInternalUpsertProfile", 0);
        // HttpClient httpClient = new HttpClient();
        // HttpClientHandler httpClientHandler = new HttpClientHandler();
        // var apiInstance = new ProfilesServiceApi(httpClient, config, httpClientHandler);

        // // Profile Upsert Request
        // var upsertRequest = new UpsertProfileRequest(profile: body.Profile);
        // result = apiInstance.ProfilesInternalUpsertProfile(userParams.UserId, "internal", upsertRequest);
        return Ok(new SuccessResponseSchema
        {
          Api = "UpdateUserProfile",
          AuthType = gatewayHeader ?? "N/A",
          HeaderUserId = userIdHeader ?? "N/A",
          PathUserId = userParams.UserId,
          Message = result.ToString()
        });
      }
      catch (ApiException e)
      {
        return BadRequest(new ErrorResponseSchema
        {
          ErrorMessage = e.Message
        });
      }

    }
  }
}
